crud/batch_upsert.lua

```diff
@@ -0,0 +1,257 @@
+        end
+
+        box.commit()
+
+        return inserted_tuples, errs
+    end
+
+    box.commit()
+
+    return inserted_tuples
+end
+
+function batch_upsert.init()
+    _G._crud.batch_upsert_on_storage = batch_upsert_on_storage
+end
+
+-- returns result, err, need_reload
+-- need_reload indicates if reloading schema could help
+-- see crud.common.schema.wrap_func_reload()
+local function call_batch_upsert_on_router(space_name, original_tuples, user_operations, opts)
+    dev_checks('string', 'table', 'table', {
+        timeout = '?number',
+        fields = '?table',
+        add_space_schema_hash = '?boolean',
+        stop_on_error = '?boolean',
+        rollback_on_error = '?boolean',
+    })
+
+    opts = opts or {}
+
+    local space = utils.get_space(space_name, vshard.router.routeall())
+    if space == nil then
+        return nil, {BatchUpsertError:new("Space %q doesn't exist", space_name)}, true
+    end
+
+    local tuples = table.deepcopy(original_tuples)
+
+    local space_format = space:format()
+    local operations = user_operations
+    if not utils.tarantool_supports_fieldpaths() then
+        operations = {}
+        for _, operations_for_tuple in ipairs(user_operations) do
+            local converted_operations, err = utils.convert_operations(operations_for_tuple, space_format)
+            if err ~= nil then
+                return nil, {BatchUpsertError:new("Wrong operations are specified: %s", err)}, true
+            end
+            table.insert(operations, converted_operations)
+        end
+    end
+
+    local batch_upsert_on_storage_opts = {
+        add_space_schema_hash = opts.add_space_schema_hash,
+        stop_on_error = opts.stop_on_error,
+        rollback_on_error = opts.rollback_on_error,
+    }
+
+    local iter, err = BatchUpsertIterator:new({
+        tuples = tuples,
+        space = space,
+        operations = operations,
+        execute_on_storage_opts = batch_upsert_on_storage_opts,
+    })
+    if err ~= nil then
+        return nil, {err}, const.NEED_SCHEMA_RELOAD
+    end
+
+    local postprocessor = BatchPostprocessor:new()
+
+    local rows, errs = call.map(BATCH_UPSERT_FUNC_NAME, nil, {
+        timeout = opts.timeout,
+        mode = 'write',
+        iter = iter,
+        postprocessor = postprocessor,
+    })
+
+    if next(rows) == nil then
+        return nil, errs
+    end
+
+    local res, err = utils.format_result(rows, space, opts.fields)
+    if err ~= nil then
+        return nil, {err}
+    end
+
+    return res, errs
+end
+
+--- Batch update or insert tuples to the specified space
+--
+-- @function tuples_batch
+--
+-- @param string space_name
+--  A space name
+--
+-- @param table tuples
+--  Tuples
+--
+-- @tparam ?table opts
+--  Options of batch_upsert.tuples_batch
+--
+-- @return[1] tuples
+-- @treturn[2] nil
+-- @treturn[2] table of tables Error description
+
+function batch_upsert.tuples_batch(space_name, tuples, user_operations, opts)
```

Comment:
I understand why you have chosen this semantic, but in seems inconvenient. Let's take a batch of ten records:
```lua
crud.upsert_many('customers',
    {
        {1, 'Name1', 23},
        {2, 'Name2', 41},
        {3, 'Name3', 72},
        {4, 'Name4', 36},
        {5, 'Name5', 85},
        {6, 'Name6', 65},
        {7, 'Name7', 56},
        {8, 'Name8', 25},
        {9, 'Name9', 13},
        {10, 'Name10', 11},
    },
    {
        {'+', 3, 1},
        {'=', 4, 'Y'},
        {'-', 3, 1},
        {'+', 3, 1},
        {'=', 4, 'Y'},
        {'-', 3, 1},
        {'+', 3, 1},
        {'=', 4, 'Y'},
        {'-', 3, 1},
        {'+', 3, 1},
    },
    { rollback_on_error = true }
)
```

It is inconvenient to understand what operations are correspond to what tuple if there are more than ~5 tuples. It seems more logical to have
```lua
crud.upsert_many('customers',
    {
        {{1, 'Name1', 23}, {'+', 3, 1}},
        {{2, 'Name2', 41}, {'=', 4, 'Y'}},
        {{3, 'Name3', 72}, {'-', 3, 1}},
        {{4, 'Name4', 36}, {'+', 3, 1}},
        {{5, 'Name5', 85}, {'=', 4, 'Y'}},
        {{6, 'Name6', 65}, {'-', 3, 1}},
        {{7, 'Name7', 56}, {'+', 3, 1}},
        {{8, 'Name8', 25}, {'=', 4, 'Y'}},
        {{9, 'Name9', 13}, {'-', 3, 1}},
        {{10, 'Name10', 11},  {'+', 3, 1}}
    },
    { rollback_on_error = true }
)
```
Also it would be more convenient to build such batches in cycle. Since this suggestion is a critical API change, I think it's worth discussing it with @Totktonada . Maybe I am wrong and it's okay to leave it as is (to be honest, I have never used an upsert in applications and don't have much idea about how batch upserts with different operations can be used).

Reply:
I agree that it seems more natural to don't split upsert operations into the array of tuples and the array of update operations, because the operation is the composition a tuple and the operations.

Reply:
BTW, `upsert` supports several operations just like `update`.

Reply:
I agree that the API proposed by @DifferentialOrange is better, but changing API and tests will take a lot of time, so I suggest resolve this in a separate task.

Reply:
As I see from the README, the upsert_many API was changed.